### PR TITLE
Fix Number.ParseNumber to not assume '\0' at the end of a span

### PR DESCRIPTION
**Pull request by stephentoub**
This routine was written for parsing strings, which are implicitly null-terminated, and it doesn't factor in string length but instead uses tricks to exit loops when the next character is null.  Now that the routine is also used for spans, this is very problematic, as spans need not be null terminated, and generally aren't when they represent slices, and expecting a null termination like this can result in walking off the end of valid memory.

I would like to see all of this code rewritten to use span.  In the interim, though, as a short-term fix I've changed all dereferences of the current position to compare against the length of the span (or, rather, a pointer to the end), and pretend that a null terminator was found if we've hit the end.

Contributes to https://github.com/dotnet/corefx/issues/29343
cc: @jkotas, @danmosemsft

### CODE REVIEW

**Comment by danmosemsft** [comment]
In general, it is hard to prove the code in this file is safe by reading it. For example NumberBuffer is assumed to be null terminated above. And it seems it always is, but this is from several frames up, where the `ReadOnlySpan<char>` always originates from a `string` before it is parsed.

File: src/mscorlib/shared/System/Number.Parsing.cs [hunk 1]
```diff
@@ -917,8 +921,13 @@ private static bool TrailingZeros(ReadOnlySpan<char> s, int index)
             // We only hurt the failure case
             // This fix is for French or Kazakh cultures. Since a user cannot type 0xA0 as a
             // space character we use 0x20 space character instead to mean the same.
-            while (*p == *str || (*str == '\u00a0' && *p == '\u0020'))
+            while (true)
             {
+                char cp = p < pEnd ? *p : '\0';
+                if (cp != *str && (*str != '\u00a0' || cp != '\u0020'))
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/mscorlib/shared/System/Number.Parsing.cs]: Ok.

**Comment by danmosemsft** [comment]
>I would like to see all of this code rewritten to use span.

@stephentoub do we need an issue to track this? I can't find one.

**Comment by stephentoub** [comment]
I don't believe we have one.

**Pull request merged**